### PR TITLE
Refactor: cleanup for release

### PR DESCRIPTION
**Pull request by amusingaxl**
Get this repo up-to-date with the most recent engineering practices, clean it up and get it ready for release.

### CODE REVIEW

**Review by oddaf — DISMISSED** [review]

**Review by 0xdecr1pto — APPROVED** [review]

File: .gitignore [hunk 1]
```diff
@@ -2,10 +2,16 @@
...
-# Ignores development broadcast logs
-!/broadcast
-/broadcast/*/31337/
-/broadcast/**/dry-run/
+# Ignores broadcast logs
+broadcast/
+
+# Ignores script config
+script/input/**/*.json
+!script/input/**/template-*.json
```
[the remark this answers is not in the document]
Comment (amusingaxl) [reply .gitignore:10]: Only files starting with `template-` are committed. The actual input files are derived from them.
This pattern is derived from `dss-direct-deposit`'s own [`.gitignore`](https://github.com/makerdao/dss-direct-deposit/blob/5684cd86f28f7dcd5e0528e5328fee48ba8ff964/.gitignore#L5-L7).
The deployment script uses `ScriptTools`, which requires a specific directory structure.
We've been using the same pattern is the recent repos we've been working on.
Since this is just a configuration file, it seems simpler to just copy and paste it instead of have it tailored on a per-repo basis.

File: foundry.toml [hunk 2]
```diff
@@ -1,6 +1,16 @@
...
+solc = '0.8.16'
+optimizer = true
 
-# See more config options https://github.com/foundry-rs/foundry/tree/master/config
\ No newline at end of file
+fs_permissions = [
+    { access = "read", path = "./out/" },
+    { access = "read", path = "./script/input/" },
+    { access = "read-write", path = "./script/output/" }
+]
```
[the remark this answers is not in the document]
Comment (amusingaxl) [reply foundry.toml:13]: Same as above applies: `ScriptTools` require this specific directory structure. Even though `script/input/` is not used, `script/output/` will be once the code is deployed.

File: README.md [hunk 3]
```diff
@@ -1,48 +1,68 @@
...
-Protego contains a factory to permissionlessly create a spell that is able to drop a single plan in the [pause](https://github.com/dapphub/ds-pause).
+If there is sufficient ecosystem interest to cancel (`drop`) a particular scheduled set of actions (the target `plan`),
+Protego contains a factory to permissionlessly create a spell ("drop spell"), which &ndash; upon being given the hat
+&ndash; is able to drop the targeted plan in `MCD_PAUSE`.
 
-* `function deploy(DSSpellLike _spell) external returns (address)`
+There are two possibilities:
 
-Used to deploy a single spell that can be elected to a privileged position. This requires that the `_spell` address parameter conforms to a MakerDAO conformant [spell](https://github.com/makerdao/spells-mainnet)
+#### 1. `deploy(DssSpellLike _spell)(address)`
```
[the remark this answers is not in the document]
Comment (amusingaxl) [reply README.md:26]: There's a case where a conforming spell might need to be dropped. For instance: some last minute legal issue that was found or a bug in the spell.

File: script/dependencies/ProtegoDeploy.sol [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+
+import {ScriptTools} from "dss-test/ScriptTools.sol";
+import {Protego} from "src/Protego.sol";
+import {ProtegoInstance} from "./ProtegoInstance.sol";
+
+struct ProtegoDeployParams {
+    address pause;
+}
+
+library ProtegoDeploy {
```
[the remark this answers is not in the document]
Comment (amusingaxl) [reply script/dependencies/ProtegoDeploy.sol:26]: In this case, it might look like an overkill because it's very simple, but this approach makes it easier to test deploy/init scripts instead of writing everything as a Foundry script.
So the answer is that we want to be consistent with other repos. Maybe to keep consistency we might want to write some tests, even if it's trivial.
You can for instance check [`dss-lite-psm`](https://github.com/makerdao/dss-lite-psm/tree/341c533d604fb56be69a847d74e8b0a523119c9e/script/dependencies) for reference.

File: src/Protego.t.sol [hunk 5]
```diff
@@ -65,337 +46,239 @@ interface DSTokenLike {
...
     address badSpell;
     address action;
     bytes32 tag;
-    bytes   sig;
+    bytes sig;
     uint256 eta;
 }
 
-contract BadAction {
+contract MaliciousSpell {
```
[the remark this answers is not in the document]
Comment (amusingaxl) [reply src/Protego.t.sol:57]: I'm gonna rearrange the contracts the way you're suggesting, however I don't get this part:

> Make NonConformantSpell define `action`, `tag`, `sig`, `eta` in way that is non-conformant / malicious

It's already doing that. `MaliciousSpell` allows anyone to `plot` any `action` + `tag` + `sig` + `eta` combination.

File: README.md [hunk 6]
```diff
@@ -1,48 +1,68 @@
...
+  permissionlessly executed.
+- Plans in `MCD_PAUSE` are [identified by a unique 32 byte hash](https://github.com/makerdao/ds-pause/blob/master/src/pause.sol#L99).
+- A conformant spell here is one where the values returned by `action()`, `tag()`, `sig()` and `eta()` are equal to the
+  corresponding `plan`'s `usr`, `tag`, `fax` and `eta` values respectively. Bad actors could potentially manipulate the
+  return values of these functions such that they are not equal, which would result in a non-conformant spell.
 
 ## Usage
 
-### Drop an individual plan
+### Create a single drop spell
```
[the remark this answers is not in the document]
Comment (amusingaxl) [reply README.md:18]: > I find it necessary to have a very clear instructions (IF/ELSE, step-by-step) explaining which of the 4 functions to use when and in which order.

Fully agree. Part of the deliverable is to provide a long-written post and also train governance facilitators on how to use the tool. We're just not sure the README is the best way to put the content verbatim. Maybe we'll have a link here to these documents once it's finished.

> As the next step, we might even consider writing a script that evaluates currently plotted spells, then fetches required parameters for the specific call and executes it.

Not sure a script would be good enough in this case. There probably needs to be a UI for that and such feature be included there, however this is out of the scope of this repo.

File: src/Protego.sol
[the remark this answers is not in the document]
Comment (amusingaxl) [reply src/Protego.sol:1]: > I'm not sure why do we need to have a custom drop function (for the deployed version) instead of conforming to the standard [DssAction](https://github.com/makerdao/dss-exec-lib/blob/master/src/DssAction.sol) interface (using schedule to drop the spell as in emergency) [...]

`drop()` exists only on `Protego` itself. On `DropSpell` we use `cast()`, which is part of the default interface. However, I see this can be confusing, as in regular spells we need to `schedule()` before `cast()`. In this case, we could move the `pause.drop()` call to `schedule` and make `cast()` an empty function to keep consistency.

> I don't think the two use cases (deploying DropSpell vs lifting the contract itself) have to be part of the same contract as it adds confusion and additional surface area.

The counter-argument would be that there is one less moving part to keep track of. In case of emergencies, people will know `Protego` is the go-to place. If we split the functionality in 2 different contracts, the cognitive overhead would be slightly increased.

> Documentation, design and tests currently does not account for the aftermath: the logic of dropping the protego itself after it has been lifted. Is this a loophole or can a separate spell be lifted in order to disable it?

Agreed that this needs to be better documented. `Protego` as a "drop any spell" facility only works as long as it has the `hat`. To disable it, governance just need to lift a different spell.

File: README.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (SidestreamColdMelon) [reply README.md:18]: My main point is: the choice to use one of the four methods could be included in the scope of the repo specifically for it to be audited. This can be done the simplest way possible, for example, using a `s.sol` file or a js-based script

File: README.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (SidestreamColdMelon) [reply README.md:26]: I understand that there are conformant spells that might need to be dropped. I rather question why not only have `deploy(address _usr, bytes32 _tag, bytes memory _fax, uint256 _eta)(address)` as it can be used to drop both types of spells and more importantly does not require expertise to decide if the spell is conformant. You can have a script along the repo to fetch all plotted spells and parse required parameters for calling this function no matter if the spell is conformant or not.

File: src/Protego.t.sol (hunk 5 above)
[the remark this answers is not in the document]
Comment (SidestreamColdMelon) [reply src/Protego.t.sol]: I meant that the `NonConformantSpell` can actually try to be malicious: it can define `action`, `tag`, `sig`, `eta` not as public _variables_ [like ConformantSpell does](https://github.com/makerdao/protego/blob/9eeb52c046e98fa2f377d480e403fccf13662872/src/test/ConformingSpell.sol#L41-L45), but as public _methods_ with custom malicious logic. For example: `tag` can return a different value when called by the pause compared to when called by protego. Meaning that it would still be conformant interface-wise, but not conformant logic-wise: making it callable via `deploy(DssSpellLike _spell)(address)` which would then deploy incorrect DropSpell.

File: README.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (SidestreamColdMelon) [reply README.md:26]: See [the comment](https://github.com/makerdao/protego/pull/2#discussion_r1600335613) for a scenario when calling `deploy(DssSpellLike _spell)(address)` could be weaponised

File: src/Protego.sol
[the remark this answers is not in the document]
Comment (amusingaxl) [reply src/Protego.sol:1]: It could be really anything: `address(0)`, an old spell that has already been cast, etc.
In either case, I'll include it in the test for the happy path.

File: src/Protego.t.sol (hunk 5 above)
[the remark this answers is not in the document]
Comment (amusingaxl) [reply src/Protego.t.sol]: A scenario like that cannot be technically prevented from a code perspective, it is more like a process issue.

I understand what you are saying, but IMO it does not help to "prove" the correctness of the implementation, which is the main goal of the tests here.
It should be clear that for non-conforming spells the version with the parameters should be used. This applies even if a non-conforming spell looks like a conforming one.

File: src/Protego.sol
[the remark this answers is not in the document]
Comment (amusingaxl) [reply src/Protego.sol:1]: > I think using schedule() + having `done()` [...]

Not sure that we want to have state for `done` here. In a situation where there is a potential governance conflict, we could have the same bad spell being scheduled more than once. That would require the ones counter-acting it to deploy another drop spell.

I don't know the specifics about the implementation of the bot, but IMO it would be better to have it calling a no-op `cast()` indefinitely until the hat changes than requiring the deployment of a new spell.

File: README.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (amusingaxl) [reply README.md:18]: Just to clarify: when you mention the "four methods", you are referring to:

1. Deploy a drop spell with a spell address (for conforming spells)
2. Deploy a drop spell from plot params (for non-conforming spells)
3. Call `Protego.drop()` with a spell address (for conforming spells)
4. Call `Protego.drop()` with plot params (for non-conforming spells)

Correct?

Do you mean we should include all flows into scripts for auditing?
What kind of additional value do you expect this to bring given that those flows are already present in the test files?

[the remark this answers is not in the document]
Comment (SidestreamColdMelon) [reply README.md:18]: > Correct?

Correct

> Do you mean we should include all flows into scripts for auditing? What kind of additional value do you expect this to bring, given that those flows are already present in the test files?

Current tests only check listed flows separately, isolated from each other. My proposal is to add usage instruction or (if possible) a script that would explain "how to choose between 4 methods" or directly execute a specific method (e.g. if there are multiple spells plotted and not executed). I propose this because this decision (no matter if described in words or in code) is security-critical (e.g. if a wrong method is used it might not drop the spell) and time-sensitive (highly likely there wouldn't be time to deploy another dropspell as voting also takes time) and therefore would highly benefit from external audit.

File: src/Protego.t.sol (hunk 5 above)
[the remark this answers is not in the document]
Comment (SidestreamColdMelon) [reply src/Protego.t.sol]: > cannot be technically prevented from a code perspective

Design-wise, I think if we don't even have `deploy(DsSpellLike _spell)(address)` and `drop(DsSpellLike _spell)` methods and instead have a script that parses required data from the original `plot` transaction it would be decrease probability of mistakes. Of course someone could still just trust the spell, but it would be less convenient.

> IMO it does not help to "prove" the correctness of the implementation, which is the main goal of the tests here.

I understand that theoretically it should not make a difference, but I personally usually rather make a test more realistic than try to imagine all edge cases that less realistic test leaves out. So I can agree that it's an opinionated suggestion

[the remark this answers is not in the document]
Comment (amusingaxl) [reply src/Protego.t.sol]: Hmm, I think I get your point. If we're talking about a conforming spell, it most likely was written by one of the teams in the spell roster, so they could easily assist governance facilitators figuring out the parameters. Maybe it makes sense to reduce the interaction surface by removing the redundant methods.

There are some things to consider, however:

> [...] instead have a script that parses required data from the original plot transaction it would be decrease probability of mistakes [...]

Who would be the end users of such script?
We want Protego to be usable regardless of the availability of engineering teams.
I guess we most likely need a UI instead of a script.

File: src/Protego.sol
[the remark this answers is not in the document]
Comment (SidestreamColdMelon) [reply src/Protego.sol:1]: > In a situation where there is a potential governance conflict, we could have the same bad spell being scheduled more than once.

I actually proposed to have `done()` as a method, not variable. It can return true as soon as no matching plan is found. So basically act as inverted [`planned()(bool)`](https://github.com/makerdao/protego/blob/5d05b7f515b88e07d2073fba83c04047ca814a1a/src/EmergencyDropSpell.sol#L85-L87). In other words be one method more compliant with [DssExec](https://github.com/makerdao/dss-exec-lib/blob/master/src/DssExec.sol).

> I don't know the specifics about the implementation of the bot

FYI, the source code is actually available here https://github.com/makerdao/chief-keeper

File: src/Protego.t.sol (hunk 5 above)
[the remark this answers is not in the document]
Comment (SidestreamColdMelon) [reply src/Protego.t.sol]: > Who would be the end users of such script?
We want Protego to be usable regardless of the availability of engineering teams.
I guess we most likely need a UI instead of a script.

I agree that the end goal is enabling non-engineers to run it. But my other point is to include the usage into the scope of the audit. Also, nothing stops you from writing a simple js script that is then directly importable as a dependency into the UI

File: src/Protego.sol [hunk 7]
```diff
@@ -1,116 +1,129 @@
...
-    //       drop any id from the pause.
-    //       This function is expected to revert if it does not have the
-    //       authority to perform this function.
+    /// @notice Permissionlessly drop anything.
+    /// @dev In some cases, due to a governance attack or other unforseen causes, it may be necessary to block any spell
+    ///      that is entered into the pause proxy.
+    ///      In this extreme case, the system can be protected during the pause delay by lifting the Protego contract to
+    ///      the hat role, which will allow any user to permissionlessly drop any id from the pause.
+    ///      This function is expected to revert if it does not have the authority to perform this function.
+    /// @param _usr The address lifted to the hat.
```
Comment (SidestreamColdMelon) [inline src/Protego.sol:110]: I think using `The address lifted to the hat` to explain `_usr` is misleading as `drop` could only be called when the protego itself is lifted to the hat. So by that explanation, it's the protego address. It might be better to use less time-dependent definition, e.g. `Address of the scheduled spell` or `Address of the plotted spell`

File: src/Protego.sol
[the remark this answers is not in the document]
Comment (amusingaxl) [reply src/Protego.sol:1]: > I actually proposed to have `done()` as a method, not variable. It can return true as soon as no matching plan is found. So basically act as inverted [`planned()(bool)`](https://github.com/makerdao/protego/blob/5d05b7f515b88e07d2073fba83c04047ca814a1a/src/EmergencyDropSpell.sol#L85-L87). In other words be one method more compliant with [DssExec](https://github.com/makerdao/dss-exec-lib/blob/master/src/DssExec.sol).

Fair enough. In this case, we have a revolving `done()`, that will return `false` again if the same spell is somehow plotted again.

File: src/Protego.sol (hunk 7 above)
Comment (amusingaxl) [reply src/Protego.sol]: Updated.

File: src/Protego.t.sol (hunk 5 above)
[the remark this answers is not in the document]
Comment (amusingaxl) [reply src/Protego.t.sol]: Agreed. However, could we do that in a follow-up PR?

File: src/Protego.sol
[the remark this answers is not in the document]
Comment (amusingaxl) [reply src/Protego.sol:1]: Code is updated to take that suggestion into account. Good stuf 💯 .

File: README.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (amusingaxl) [reply README.md]: Agreed, it will make it more difficult instead of easier for users.
Code was updated.

File: README.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (amusingaxl) [reply README.md]: I'm hoping that the documentation is now clear enough, since there are only 2 methods left.
Maybe there should be something written in a "decision tree" format, but not sure if it's strictly necessary.

File: script/dependencies/ProtegoDeploy.sol (hunk 4 above)
[the remark this answers is not in the document]
Comment (SidestreamColdMelon) [reply script/dependencies/ProtegoDeploy.sol:26]: For testing, Foundry actually recommends to directly ["test your scripts"](https://book.getfoundry.sh/tutorials/best-practices#:~:text=on%20your%20project.-,Test%20your%20scripts.,-Unit%20test%20them). But I think the main reason to use a library from a script is more than just code/organisation. From my understanding, a library would actually be deployed first in order to be used in a script, but then all of the operations in that library would be executed in a single transaction compared to a `s.sol` script where every operation is a separate transaction. In other words, it is to prevent front-running possibility. Although the above doesn't make any difference in this particular case, I agree with your point, let's keep it for consistency

File: src/Protego.t.sol [hunk 8]
```diff
@@ -65,337 +47,157 @@ interface DSTokenLike {
...
 
-            assertTrue(protego.planned(badSpells[i].action, badSpells[i].tag, badSpells[i].sig, badSpells[i].eta));
+            assertFalse(protego.planned(badSpells[i].action, badSpells[i].tag, badSpells[i].sig, badSpells[i].eta));
         }
 
-        vote(address(protego));
+        // After Protego loses the hat, it can no longer drop spells
+        _vote(address(0));
+        vm.expectRevert();
+        protego.drop(badSpells[0].action, badSpells[0].tag, badSpells[0].sig, badSpells[0].eta);
```
Comment (SidestreamColdMelon) [inline src/Protego.t.sol:183]: Is it `ds-pause-proxy-unauthorized` with which it fails? Then, could you please specify it in `vm.expectRevert("ds-pause-proxy-unauthorized");`

File: src/EmergencyDropSpell.sol [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+
+    /// @notice Returns whether the original spell has been planned or not.
+    function planned() public view returns (bool) {
+        return protego.planned(action, tag, sig, eta);
+    }
+
+    /// @notice Drops the original spell.
+    function schedule() external {
+        pause.drop(action, tag, sig, eta);
+    }
```
Comment (SidestreamColdMelon) [inline src/EmergencyDropSpell.sol:91]: Should we also emit `Drop` event here for consistency (since it can be executed multiple times)? Or maybe even call protego.drop so that events accumulate under the same address

File: src/test/NonConformingSpell.sol [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+        pause = DsPauseLike(pause_);
+    }
+
+    function plot(address usr, bytes32 tag, bytes memory fax, uint256 eta) public {
+        pause.plot(usr, tag, fax, eta);
+    }
+
+    function drop(address usr, bytes32 tag, bytes memory fax, uint256 eta) public {
+        pause.drop(usr, tag, fax, eta);
+    }
```
Comment (SidestreamColdMelon) [inline src/test/NonConformingSpell.sol:37]: Why did you decide to add `drop` to the `NonConformingSpell`? I don't see it's being used anywhere

File: src/EmergencyDropSpell.sol [hunk 11]
```diff
@@ -0,0 +1,106 @@
...
+
+    /// @param _protego The Protego factory that deployed the spell.
+    /// @param _pause The MCD_PAUSE instance.
+    /// @param _usr The original spell action address.
+    /// @param _tag The original spell action tag (i.e.: `extcodehash`).
+    /// @param _fax The original spell encoded call.
+    /// @param _eta The original spell expiry time.
+    constructor(address _protego, address _pause, address _usr, bytes32 _tag, bytes memory _fax, uint256 _eta) {
+        protego = ProtegoLike(_protego);
+        pause = DsPauseLike(_pause);
```
[the remark this answers is not in the document]
Comment (amusingaxl) [reply src/EmergencyDropSpell.sol:71]: This would mean one additional `CALL` in the transaction to deploy it. However, gas costs are not exactly the biggest priority here, so I see no harm in pulling if from `protego`.

File: src/EmergencyDropSpell.sol (hunk 9 above)
Comment (amusingaxl) [reply src/EmergencyDropSpell.sol:91]: `protego.drop` is only possible if `Protego` has the hat.

File: src/test/NonConformingSpell.sol (hunk 10 above)
Comment (amusingaxl) [reply src/test/NonConformingSpell.sol:37]: Just for interface completeness sake. I can remove it.

File: src/Protego.t.sol (hunk 8 above)
Comment (amusingaxl) [reply src/Protego.t.sol:183]: Do you mean `ds-auth-unauthorized`? I'll add that for explicitness.

File: src/EmergencyDropSpell.sol (hunk 9 above)
Comment (amusingaxl) [reply src/EmergencyDropSpell.sol]: Please check now. I think I found a good solution for the naming issue.

File: src/Protego.t.sol (hunk 8 above)
Comment (amusingaxl) [reply src/Protego.t.sol]: Done.

File: src/EmergencyDropSpell.sol (hunk 11 above)
[the remark this answers is not in the document]
Comment (amusingaxl) [reply src/EmergencyDropSpell.sol]: Done.

File: src/EmergencyDropSpell.sol [hunk 12]
```diff
@@ -0,0 +1,106 @@
...
+    function schedule() external;
+    function sig() external view returns (bytes memory);
+    function tag() external view returns (bytes32);
+}
+
+/// @title A spell that drops a plan from `MCD_PAUSE` when is cast.
+contract EmergencyDropSpell is EmergencySpellLike {
+    bool public constant officeHours = false;
+    uint256 public constant expiration = type(uint256).max;
+    address public constant log = address(0);
```
[the remark this answers is not in the document]
Comment (amusingaxl) [reply src/EmergencyDropSpell.sol]: Done.

File: src/EmergencyDropSpell.sol (hunk 9 above)
Comment (SidestreamColdMelon) [reply src/EmergencyDropSpell.sol]: > `protego.drop` is only possible if `Protego` has the hat. Also not sure if we should call this event `Drop` or `Schedule` 😅

You can do delegatecall, but I don't know if it's a good practice to use delegate call for the sake of having events in one place.

**Review by SidestreamColdMelon — APPROVED** [review]

File: src/EmergencyDropSpell.sol (hunk 9 above)
Comment (amusingaxl) [reply src/EmergencyDropSpell.sol]: Yeah, in this case it would work because there's no state involved, however I believe this would not land well with the existing security practices. A spell with `delegatecall` would cause an aneurysm in some people 😅.

File: script/dependencies/ProtegoDeploy.sol (hunk 4 above)
[the remark this answers is not in the document]
Comment (amusingaxl) [reply script/dependencies/ProtegoDeploy.sol:26]: > For testing, Foundry actually recommends to directly ["test your scripts"](https://book.getfoundry.sh/tutorials/best-practices#:~:text=on%20your%20project.-,Test%20your%20scripts.,-Unit%20test%20them).

Yeah, but honestly it feels very cumbersome to test scripts because off all the context the `Script` base contract brings.

The script deals with "all things external", such as reading from config files/environment, while the library remains agnostic of configuration.

> [...] a library would actually be deployed first in order to be used in a script [...]

Actually, we don't deploy such libraries, they are imported as-is by spells. See [this example](https://github.com/makerdao/spells-mainnet/blob/master/archive/2023-07-14-DssSpell/dependencies/dss-flappers/FlapperInit.sol).

> [...] but then all of the operations in that library would be executed in a single transaction compared to a s.sol script where every operation is a separate transaction.

Since the library has only internal functions, the code is actually inlined by the compiler, so we actually have multiple transactions being executed when imported by a script.

[the remark this answers is not in the document]
Comment (SidestreamColdMelon) [reply script/dependencies/ProtegoDeploy.sol:26]: > Since the library has only internal functions, the code is actually inlined by the compiler

Thanks for pointing that out, I totally missed that fact.

**Review by 0xp3th1um — APPROVED** [review]

**Review by oddaf — APPROVED** [review]
